Lots of tweaks

nichogenius opened the pull request:
-- Added a 3rd pattern file, patterns_iraw.txt.  Exactly the same as patterns_raw.txt except it does a case insensitive search.  Useful for finding phishers using words like phishtank etc

-- Added pattern comment and whitespace handling.  We can now add comments and blank lines to the pattern files to make them easier to document and format.  Started sorting but I don't know what many of the patterns are picking up.

-- Changed the output format.  The md5 checksum now comes before the path as that column is fixed width... might not be as pretty if you use a narrow terminal window.  Also added color to the md5 and the pattern matched to make it easier to pick out the individual parts.  Added '#' characters into the output to make the output bash safe in the event of accidental paste dumps into the terminal as '#' is the comment symbol for bash.  I'm still not completely happy with the output.

-- Created more flag options so there is a short and long form of each flag.  ie --directory or -d, --ignore or -i, --hide-ok or -k.

-- Added verbose option which will continue scanning the same file even after a first match is found... useful for finding what patterns are the most common or additional datapoints to classify a suspicious file.

-- Added a python script called b64p.py which takes an input string and returns the 3 base64 string equivalents.  Usage: ./b64p.py 'preg_match'

-- Added php_keywords.txt.  This is a file containing a list of PHP 7 keywords and their base64 translations.  It is formatted in such a way as to be able to be used as a pattern file as a replacement for patterns_raw.txt.  Useful to scan known malware to find new patterns to look for.

-- Added php_functions.txt.  Similar to php_keywords.txt, but is specific to php 7 functions.  Very large and very slow to scan with, but it's a good reference file and useful to scan known malware with.

-- Sorted the Whitelist file first by path, then by md5 checksum

-- Modified the part of the code responsible for reading in pattern files.  Read each pattern file in once instead of reading in each pattern file again each time a new .php file is scanned.  Should boost performance marginally.

nichogenius committed:
Sorted this file by column 2

cat whitelist.txt | sort -k 2 | less
No reason this shouldn't be sorted perfectly to keep like files together.
No white list rules changed... just plain sorting.

nichogenius committed:
Two-key sorting is better

cat php-malware-scanner-master/whitelist.txt | sort -k 2,2 -k 1,1 | less
More of an OCD thing than anything, but might as well sort primarily by file path, secondarily by hash value.

nichogenius committed:
Whitespace and Comment handling for pattern files

The pattern files are large and complex enough to justify some whitespace and comments to explain what each entry is.

Added logic to check if the line is empty or if the first character is equal to '#' before using it as a pattern.  Simply skips over empty and commented lines.

nichogenius committed:
Added generic base64 long string pattern

Added some comments to the file and added a generic base64 string regex for long base64 strings.

nichogenius committed:
Organizing, categorizing and prioritizing patterns

There's enough raw patterns in here to justify organizing the file.
Now that whitespace and comments are supported, I've been dividing it into sections

More critical problems should be near the top as I would rather the script identify a file as a backdoor instead of as a spammer.

I don't know the history behind a lot of these or the implication of the code, so I'm sure I mis-categorized many.  There are also many that I have not done yet.

nichogenius committed:
Bug fix for last commit

comment and empty line filters were in the wrong place

nichogenius committed:
using strpos instead of substr_count

don't know if it's faster, but I don't see a reason to count the number of times a line exists in a file for our use case.

nichogenius committed:
New file for case-insensitive search of strings

New file for case-insensitive search of strings
patterns_iraw.txt

nichogenius committed:
Added case-insensitive search logic

Added case-insensitive search logic

will search patterns_raw.txt, patterns_iraw.txt and patterns_re.txt

nichogenius committed:
Added Security service providers

Phishers often block these web-security providers... worth checking for if they don't create many false positives.

nichogenius committed:
Add files via upload

This is a file of base64 patterns that represent strings that would be present if any of the functions in php7 were encoded to base64.  I'll probably add structure later by grouping them with their plain text translation.

This file is useful to swap out with patterns_raw.txt to gain additional insights into other strings to search for in base64.

nichogenius committed:
Removed some nasty false positive patterns

Removed some nasty false positive patterns

nichogenius committed:
Added functions, dropped Spammer.

Spammers gives false positives.  added a couple more php functions in base64

nichogenius committed:
Removed (chr\(\d+\)\.){4,}

Too many false positives

nichogenius committed:
Adding base_convert as a test

base_convert seems like a rare enough function... trying it out to see if it generates many fp's.

nichogenius committed:
bad idea... lots of fps

bad idea... lots of fps
removing base_convert

nichogenius committed:
ovh removed

ovh removed

nichogenius committed:
Uploaded an organized and commented version

This file just contains a list of internal php 7 functions (probably incomplete depending on extensions etc) and their 3 base64 fingerprints. It is designed to be used as either a pattern file to explore potential patterns that may be effective, or simply as a reference to translate between plain text php and the 3 different base64 versions.

nichogenius committed:
just adding some comments

just adding some comments

nichogenius committed:
added create_function

interesting note from the php.net manual on create_function:
Caution
This function internally performs an eval() and as such has the same security issues as eval(). Additionally it has bad performance and memory usage characteristics.

nichogenius committed:
added file_get_contents, fopen

dropped allow_url_fopen as fopen will catch those references on its own.

nichogenius committed:
shortened base64_decode to just base64

shortened the base64 fingerprints of 'base64_decode' to just 'base64'. will also catch cases of base64_encode which isn't quite so bad but still worth finding.

nichogenius committed:
Typo fix

Typo fix

nichogenius committed:
Added obfuscat

obfuscat should catch things like obfuscated, obfuscator, obfuscation.
Some online obfuscator providers leave a signature comment in the code.

nichogenius committed:
Added http://www.fopo.com.ar/

http://www.fopo.com.ar/ is a free online php obfuscator that apparently leaves comments in the code which we should be looking for.

nichogenius committed:
removed datatran

removed datatran for false positives

nichogenius committed:
obfuscat is too common, causes fp's

obfuscat is too common, causes fp's

nichogenius committed:
Added createfunction and gzinflate obfuscations

Added createfunction and gzinflate obfuscations.
also added gzinflate b64 patterns

nichogenius committed:
Found a bug in my base64 converter

Found a bug in my base64 converter

My base64 conversion script is supposed to find the maximum length string that is guaranteed to be present if the input plain text string is somewhere in the original plain text code, however there was an off by 1 error which made some patterns 1 character longer than they should have been.   Short patterns (ie 4 chars) were prone to false positives because they really were 3 character patterns which is too short to be useful.  Long patterns were likely missing results.

Should be fixed now.

nichogenius committed:
This is how I generate base64 sample patterns.

Example usage:

I want to see if a giant block of base64 code contains any references to the string 'base64'. 
The naive approach is to convert the string to it's base64 equivalent, YmFzZTY0.

There are two problems with this approach.  The first is that the string will be different depending on the position of the first character 'Y' in the input string.  Possible offents are 0 bits, 2 bits or 4 bits.  The above example only calculates the 0 bit offset.  There should be 3 separate base64 strings to look for.

The second problem is that base64 strings use a 6 bit encoding, so the characters don't align the same as 8 bit encoding.  This leads to character bleeding at the beginning and ends of a string where the string will change depending on its immediate context.  This script calculates the maximum constant string length that should be present.  Unfortunately it requires trimming characters which can often lead to very short strings.

nichogenius committed:
I was wrong... original code was correct.

I was wrong... original code was correct... there was no bug but the one in my head

nichogenius committed:
this is not a bug... just fixing a bad bug fix

this is not a bug... just fixing a bad bug fix.
there is no off by 1 error after all.

nichogenius committed:
removed mail b64, added chr b64

removed mail b64, added chr b64

mail was generating too many false positives.
chr has only one pattern that is long enough to use with any sort of reliability, but it is one that we want to look out for anyway.

nichogenius committed:
Added array_ and cslashes

Found a couple of cases where the php functions array_shift and addcslashes were used in base64 encoded malware.

Adding strings to catch any references to 'cslashes' which will catch both addcslashes and strip cslashes
Adding strings to catch any references to 'array_' which will catch about a dozen array modification functions.

nichogenius committed:
PHP Keywords in Base64, pattern file

This is a file containing a list of PHP keywords converted to base64.  It's designed to be used as a pattern file to identify common keywords used in obfuscated code.

nichogenius committed:
Renamed to match naming conventionsi

Renamed to match naming conventionsi

nichogenius committed:
Copied comments from php_functions.php

Copied comments from php_functions.php

nichogenius committed:
Adding str_, function, echo and include in base64

str_ will match 13 separate php functions, many of which can be used for string/modifcation aka obfuscation
function added to catch function defining.
echo added as it is a  common php  keyword, though experimental... may cause a of false positives
include added as it is often used to link in other malware files.

nichogenius committed:
Added encoded versions of '_' character.

Added encoded versions of '_' character.

nichogenius committed:
Added new flag options

Added a single short flag for every long flag and a single long flag for every short flag.
This now gives us 2 ways to set each flag.
Also updated the showhelp.
Dropped an unnecessary 'else' statement.

nichogenius committed:
Added long single line PHP code pattern

One common tactic is to shove all of your PHP code into a single line, often contained within its own PHP tags, and drop it into any .php file that you want.  This pattern should detect if more than 750 characters are contained within PHP tags on a single line.

nichogenius committed:
preg_match 's' flag changed to 'm'

the 's' flag tells preg_match to operate in multi-line mode.  the 'm' flag does the same, but allows line begin and ends to still be matched which is useful in some cases.

nichogenius committed:
Added Experimental Patterns

Removed anyresults.net from the base64 pattern strings.

Added base64 patterns for $_POST $_GET $_REQUEST $_COOKIE extract and GLOBALS

nichogenius committed:
Added New Malware Signatures/Fingerprints

IndoXploit
FaisaL Ahmed aka rEd X
'F'.'il'.'esMan'
FilesMan

nichogenius committed:
Several Significant changes to scan.php

- Gave each flag option a short or long option; like i:ignore or d:directory or k:hide-ok
- Added a verbose option that instructs the scan to scan a file for ALL matches and not just stop at the first one.
- Restructured the output code to allow for the verbose flag, mainly a new function printPath and where the md5 hash is computed
- Modified the output to be cleaner, checksum is printed first as it is fix-width and to make it easier to paste into the whitelist file.
- Modified the output to be 'bash safe', ie when I accidentally paste my scan results into my terminal, the '#' should make sure everything is treated as a comment.  This is in contrast to possibly attempting to execute absolute paths to potentially malicious PHP scripts and the usage of the '>' which tells the shell to write to a file.  Also enclosed each path in {} for similar purposes.
- Printing the matched string/pattern in $color... might change later depending on  preference.

nichogenius committed:
Raised Long PHP line Limit

Previous limit of 750 chars enclosed in php tags in a single line was too low... false positives were being triggered by a w3 total cache file because some guy decided to print one gigantic long message in a single line.

Raising to 1100

nichogenius committed:
Removing fopen('/etc/passwd - Redundant

fopen('/etc/passwd dropped
sub-pattern /etc/passwd exists and is completely contained within fopen('/etc/passwd

nichogenius committed:
FilesMan Needs to be a regex

FilesMan Needs to be a regex

nichogenius committed:
Verbose Bug fix and pattern loading optimization

Verbose flag was not proceeding with the next scan due to !found being set.  Added a check to see if it is verbose when it decides to do the next scan.

Patterns should be loaded once and only once.  The files aren't large so not a problem with memory, however it might impact performance if we are loading the same 3 files ever time we scan a file.

nichogenius committed:
Pattern Loading Moved To Constructor

It makes more sense to put the one time pattern load code into the constructor rather than the scan method.

nichogenius committed:
Cleaned up pattern whitespace and comment handling

checking for comments and whitespace lines in pattern files should only be done once when the patterns are loaded.
Added this code to the loadPatterns() function and removed it from the 3 scan loops.

scr34m commented:
Oh god :) lots of changes, will check on weekend, but I don't see any problem to accept, just to be sure, thanks!

nichogenius commented:
As far as detriments, I think i only dropped about 3 patterns (from the original) that were causing too many false positives for my liking.  Spammer, (chr(\d+)\.){4}, and one other I can't think of right now.

Most of the changes were adding functionality or just tidying some stuff up.

nichogenius commented:
Probably best if you try out the changes before accepting... the output formatting might be a bit too much of an overhaul.

nichogenius committed:
Bug Fixes, added time/checksum flags, organized

--Fixed a bug with the out function.  Previous updates of mine did not update all calls to the out function which I changed the parameters for.  Fixed this by replacing the out function with an 'error' function.
--Alphabetized function definitions and did some general tidying up
--Made all functions private except the constructor.
--Created parseArgs function to handle reading in options.
--Fixed a bug with 'extra-check' where htaccess and googleBot were being pushed to the pattern array each time a file was scanned. 
 This bug was created when I moved the pattern initialize code to the constructor.  Moved extra-check code with the rest of the initialize pattern calls.
--Added -no-color, -time, and -checksum flags.  I'd prefer if the output was only as spammy as the user requests.  Time should be helpful in tracing when the attack occurred and if files are related to the same hack.  Time and checksum do not display by default.  no-color flag makes it easier to dump to plain text files.

nichogenius committed:
base64 pattern updates

nichogenius committed:
removed redundant pattern

'create_function' is redundant as 'function' is already present.

nichogenius committed:
added b64 pattern for 'require'

added b64 pattern for 'require'

nichogenius committed:
Added comment lines for each regex

Added comment lines for each regex
This is to allow the -comment flag to provide either an accurate comment or no comment at all.

nichogenius committed:
large 'scan' function broken up comment flag added

scan function has been broken up into a multiple functions which should make it more modular if future scan types are needed.
--comment, -a flag was added which prints the first comment to appear prior to the matched pattern in the pattern file.

nichogenius committed:
updated flags

updated flags

checksum can be -m for md5
comment can be -c for comment

nichogenius committed:
fixed -m checksum flag

fixed -m checksum flag

nichogenius committed:
Added Comments and Documentation

Added Comments and Documentation
